### PR TITLE
[PLAT-112239] Add e2e latency metrics for Thanos receive

### PR DESCRIPTION
**Pull request by yuchen-db**
since this is a public repo, i put test screenshots in https://github.com/databricks/universe/pull/645497

### CODE REVIEW

File: pkg/receive/handler.go [hunk 1]
```diff
@@ -462,6 +472,17 @@ func newWriteResponse(seriesIDs []int, err error, er endpointReplica) writeRespo
 	}
 }
 
+func secondsSinceOldestSample(ts prompb.TimeSeries) float64 {
+	now := time.Now().UnixNano() / int64(time.Millisecond)
+	oldestTs := now
```
Comment (jnyi) [inline pkg/receive/handler.go]: don't generate now for every series, do it once in `receiveHTTP`

File: pkg/receive/handler.go [hunk 2]
```diff
@@ -606,6 +627,11 @@ func (h *Handler) receiveHTTP(w http.ResponseWriter, r *http.Request) {
 		h.writeTimeseriesTotal.WithLabelValues(strconv.Itoa(responseStatusCode), tenant).Observe(float64(stats.timeseries))
 		h.writeSamplesTotal.WithLabelValues(strconv.Itoa(responseStatusCode), tenant).Observe(float64(stats.totalSamples))
 	}
+	for _, ts := range wreq.Timeseries {
```
Comment (jnyi) [inline pkg/receive/handler.go:630]: get a `now` here just once and calculate e2e latency.

File: pkg/receive/handler.go [hunk 3]
```diff
@@ -208,6 +209,15 @@ func NewHandler(logger log.Logger, o *Options) *Handler {
 				Buckets:   []float64{10, 50, 100, 500, 1000, 5000, 10000},
 			}, []string{"code", "tenant"},
 		),
+		writeE2eLatency: promauto.With(registerer).NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: "thanos",
+				Subsystem: "receive",
+				Name:      "write_e2e_latency_seconds",
+				Help:      "The end-to-end latency of the oldest sample in write requests.",
+				Buckets:   []float64{1, 5, 10, 20, 30, 40, 50, 60, 90, 120, 300, 600, 900, 1200, 1800, 3600},
```
Comment (jnyi) [inline pkg/receive/handler.go:218]: cardinality estiamte: 16 le x 5 code x 20 tenants x 30 pods = 48k seems fine.

Comment (yuchen-db) [reply pkg/receive/handler.go:218]: i'm thinking if we can drop "code" label as failed requests can be captured by other metrics. but even if we do not drop it, the data samples will be minimal. WDYT?

**Review by christopherzli — APPROVED** [review]
LGTM

File: pkg/receive/handler.go (hunk 3 above)
Comment (christopherzli) [reply pkg/receive/handler.go:218]: why do we need latency buckets like 3600?

Comment (yuchen-db) [reply pkg/receive/handler.go:218]: In normal cases e2e latency can be ~ 10 minutes. Extending the bucket to 1 hour to (hopefully) cover future outage

**Pull request merged**